python/context_selector/context_widget.py

```diff
@@ -0,0 +1,956 @@
+                elif event.key() in [
+                    QtCore.Qt.Key_Tab,
+                    QtCore.Qt.Key_Return,
+                    QtCore.Qt.Key_Enter,
+                ]:
+                    # user hit tab/enter/return in search. go with the currently
+                    # highlighted item or the first one
+                    result = \
+                        self.ui.link_search.completer().get_current_result() or\
+                        self.ui.link_search.completer().get_first_result()
+                    if result:
+                        self._on_entity_activated(
+                            result["type"],
+                            result["id"],
+                            result["name"],
+                        )
+
+        return False
+
+    def save_recent_contexts(self):
+        """
+        Should be called by the parent widget, typically when the dialog closes,
+        to ensure the recent contexts are saved to disk when closing.
+        """
+
+        # build a list of serialized recent contexts. we grab all the QActions
+        # from the recents list ad serialize them.
+        serialized_contexts = []
+        for recent_action in self._menu_actions["Recent"]:
+            recent_context = recent_action.data()
+
+            # don't include the user credentials in the serialized context as
+            # it may cause issues with authentication when deserializing
+            serialized_context = recent_context.serialize(
+                with_user_credentials=False)
+
+            serialized_contexts.append(serialized_context)
+
+        logger.debug("Storing serialized 'Recent' contexts.")
+
+        # store the recent contexts on disk. the scope is per-project
+        self._settings.store(
+            self._settings_recent_contexts_key,
+            serialized_contexts,
+            scope=settings.UserSettings.SCOPE_PROJECT
+        )
+
+    def set_context(self, context, task_display_override=None,
+        link_display_override=None):
+        """
+        Set the context to display in the widget.
+
+        The initial display values can be overridden via the task and link
+        override args.
+
+        :param context: Toolkit Context that the widget should be set to.
+        :param str task_display_override: Override text to be displayed for the task.
+        :param str link_display_override: Override text to be displayed for the link.
+        """
+        logger.debug("Setting context to: %s" % (context,))
+
+        # clear any related tasks from the previous context
+        self._menu_actions["Related"] = []
+
+        self._context = context
+        self._show_context(
+            context,
+            task_display_override=task_display_override,
+            link_display_override=link_display_override
+        )
+
+        # ensure the new context is added to the list of recents.
+        # TODO: we should only update recents after publishing. but until we
```

Comment:
maybe remove this comment since this is outside the publisher now?

Reply:
done